### PR TITLE
Non root contaienr

### PR DESCRIPTION
**Pull request by ninjadq**
* Core container
* portal container
* jobservice
* proxy
* redis
* notary
* database

### CODE REVIEW

**Comment by cadnce** [comment]
👍 similar to changes i've just made.

However i think you need to update the following ports to 8080:
https://github.com/goharbor/harbor/blob/master/make/photon/prepare/templates/nginx/nginx.https.conf.jinja#L21 
https://github.com/goharbor/harbor/blob/master/make/photon/prepare/templates/nginx/nginx.http.conf.jinja#L20

File: make/photon/portal/nginx.conf [hunk 1]
```diff
@@ -6,8 +6,15 @@ events {
 }
```
Comment (cadnce) [inline make/photon/portal/nginx.conf]: Do you need to set `pid /tmp/nginx.pid` in the top level directive here?

File: make/photon/core/Dockerfile [hunk 2]
```diff
@@ -12,5 +12,5 @@ COPY ./make/migrations /harbor/migrations
 
 RUN chmod u+x /harbor/start.sh /harbor/harbor_core
 WORKDIR /harbor/
-	
-ENTRYPOINT ["/harbor/start.sh"]
+
```
[the remark this answers is not in the document]
Comment (ninjadq) [reply make/photon/core/Dockerfile]: Thanks for review comments, this pr is still WIP 👍

[the remark this answers is not in the document]
Comment (cadnce) [reply make/photon/core/Dockerfile]: Awesome, just keen to get a neat set of containers that don't run as root (I say I made the same set of changes prior to this as was just waiting to contribute them)

My difficulty was around how you manage the registry container as it doesn't need to be sudo'd to run except for the e2e tests (according to the comments)

File: make/photon/portal/Dockerfile [hunk 3]
```diff
@@ -10,9 +10,8 @@ WORKDIR /build_dir
 RUN cp -r /portal_src/* /build_dir \
     && ls -la \
     && apt-get update \
-    && apt-get install python-yaml \
-    && cat swagger.yaml | python -c 'import sys, yaml, json; y=yaml.load(sys.stdin.read()); print json.dumps(y)' > swagger.json \
-    && cat ./package.json \
```
Comment (reasonerjt) [inline make/photon/portal/Dockerfile]: What was the reason for `cat ./package.json`?  Do you recall?

File: make/photon/jobservice/Dockerfile [hunk 4]
```diff
@@ -1,13 +1,19 @@
...
 
-RUN mkdir /harbor/ \
-    && tdnf install sudo -y >> /dev/null\
+RUN tdnf install sudo -y >> /dev/null\
     && tdnf clean all \
-    && groupadd -r -g 10000 harbor && useradd --no-log-init -r -g 10000 -u 10000 harbor 
+    && groupadd -r -g 10000 harbor && useradd --no-log-init -r -g 10000 -u 10000 harbor
 
-COPY ./make/photon/jobservice/start.sh ./make/photon/jobservice/harbor_jobservice /harbor/
+COPY ./make/photon/jobservice/harbor_jobservice /harbor/
```
Comment (reasonerjt) [inline make/photon/jobservice/Dockerfile]: We don't need to create `/harbor/` beforehand?

File: make/photon/jobservice/Dockerfile [hunk 5]
```diff
@@ -1,13 +1,19 @@
 FROM photon:2.0
 
-RUN mkdir /harbor/ \
-    && tdnf install sudo -y >> /dev/null\
+RUN tdnf install sudo -y >> /dev/null\
```
Comment (reasonerjt) [inline make/photon/jobservice/Dockerfile]: do we still need `sudo`?

File: make/photon/portal/Dockerfile [hunk 6]
```diff
@@ -1,39 +1,44 @@
...
-COPY --from=nodeportal /build_dir/swagger.json /usr/share/nginx/html
-
-COPY make/photon/portal/nginx.conf /etc/nginx/nginx.conf
-
 STOPSIGNAL SIGQUIT
 
-HEALTHCHECK CMD curl --fail -s http://127.0.0.1 || exit 1
-
+HEALTHCHECK CMD curl --fail -s http://127.0.0.1:8080 || exit 1
+USER nginx
```
Comment (reasonerjt) [inline make/photon/portal/Dockerfile]: line 36 still calls `EXPOSE 80`?

File: make/photon/portal/nginx.conf [hunk 7]
```diff
@@ -1,13 +1,21 @@
 
-worker_processes  1;
+worker_processes auto;
```
Comment (reasonerjt) [inline make/photon/portal/nginx.conf]: What's the rationale behind this change?

File: make/photon/prepare/utils/jobservice.py [hunk 8]
```diff
@@ -18,7 +18,8 @@ def prepare_job_service(config_dict):
 
     # Job log is stored in data dir
     job_log_dir = os.path.join('/data', "job_logs")
-    prepare_config_dir(job_log_dir)
+    file_path = prepare_config_dir(job_log_dir)
```
Comment (reasonerjt) [inline make/photon/prepare/utils/jobservice.py]: this is trivial but why assign another var instead of reusing `job_log_dir`

File: make/photon/prepare/utils/nginx.py [hunk 9]
```diff
@@ -17,12 +17,17 @@
...
 def prepare_nginx(config_dict):
-    prepare_config_dir(nginx_confd_dir)
+    file_path = prepare_config_dir(nginx_confd_dir)
+    os.chown(file_path, DEFAULT_UID, DEFAULT_GID)
     render_nginx_template(config_dict)
 
 def render_nginx_template(config_dict):
     if config_dict['protocol'] == "https":
-        render_jinja(nginx_https_conf_template, nginx_conf,
+        render_jinja(
```
Comment (reasonerjt) [inline make/photon/prepare/utils/nginx.py]: does nginx conf file need `uid` `gid`?

File: make/photon/prepare/utils/nginx.py [hunk 10]
```diff
@@ -35,7 +40,9 @@ def render_nginx_template(config_dict):
     else:
         render_jinja(
             nginx_http_conf_template,
-            nginx_conf)
+            nginx_conf,
+            uid=DEFAULT_UID,
```
Comment (reasonerjt) [inline make/photon/prepare/utils/nginx.py]: does nginx conf file need uid gid?

File: make/photon/prepare/utils/notary.py [hunk 11]
```diff
@@ -72,9 +72,12 @@ def prepare_env_notary(nginx_config_dir):
 
 
     print("Copying nginx configuration file for notary")
-    shutil.copy2(
+
+    render_jinja(
         os.path.join(templates_dir, "nginx", "notary.upstream.conf.jinja"),
-        os.path.join(nginx_config_dir, "notary.upstream.conf"))
+        os.path.join(nginx_config_dir, "notary.upstream.conf"),
+        gid=DEFAULT_GID,
```
Comment (reasonerjt) [inline make/photon/prepare/utils/notary.py]: does nginx conf file need uid gid?

File: make/photon/prepare/utils/notary.py [hunk 12]
```diff
@@ -88,6 +91,8 @@ def prepare_notary(config_dict, nginx_config_dir, ssl_cert_path, ssl_cert_key_pa
     render_jinja(
         notary_server_nginx_config_template,
         os.path.join(nginx_config_dir, "notary.server.conf"),
+        gid=DEFAULT_GID,
```
Comment (reasonerjt) [inline make/photon/prepare/utils/notary.py]: does nginx conf file need uid gid?

File: make/photon/prepare/utils/redis.py [hunk 13]
```diff
@@ -0,0 +1,14 @@
+import os
+
+from g import data_dir, REDIS_UID, REDIS_GID
+from utils.misc import prepare_config_dir
+
+redis_data_path = os.path.join(data_dir, 'redis')
+
+def prepare_redis(config_dict):
+    prepare_config_dir(redis_data_path)
```
Comment (reasonerjt) [inline make/photon/prepare/utils/redis.py]: Generally after going over the changes in python, it seems better if we rename the `prepare_config_dir` to `prepare_dir` and make `uid` `gid` optional parms, so that there will be less repetitive code.

File: make/photon/redis/Dockerfile [hunk 14]
```diff
@@ -4,11 +4,12 @@ RUN tdnf install -y redis sudo
 
 VOLUME /var/lib/redis
 WORKDIR /var/lib/redis
-COPY ./make/photon/redis/docker-entrypoint.sh /usr/bin/
+COPY ./make/photon/redis/ /usr/bin/
```
Comment (reasonerjt) [inline make/photon/redis/Dockerfile]: Why copy the directory instead of the health-check script?

**Comment by reasonerjt** [comment]
Please see comments and rebase to resolve the conflict.

File: make/photon/jobservice/Dockerfile (hunk 5 above)
Comment (ninjadq) [reply make/photon/jobservice/Dockerfile]: yes, `chown`, `chmod` are included in sudo package, we need these command while build the image

File: make/photon/portal/nginx.conf (hunk 7 above)
Comment (ninjadq) [reply make/photon/portal/nginx.conf]: `auto` will improve the performance when input traffic is large.

File: make/photon/prepare/utils/jobservice.py (hunk 8 above)
Comment (ninjadq) [reply make/photon/prepare/utils/jobservice.py]: I think `prepare_config_dir` function will return a value which is the real path, although it is the same as the original one if only one arg passed.

File: make/photon/prepare/utils/nginx.py (hunk 9 above)
Comment (ninjadq) [reply make/photon/prepare/utils/nginx.py]: of course

File: make/photon/prepare/utils/nginx.py (hunk 10 above)
Comment (ninjadq) [reply make/photon/prepare/utils/nginx.py]: yes

File: make/photon/prepare/utils/notary.py (hunk 11 above)
Comment (ninjadq) [reply make/photon/prepare/utils/notary.py]: yes

File: make/photon/jobservice/Dockerfile (hunk 4 above)
Comment (ninjadq) [reply make/photon/jobservice/Dockerfile]: `COPY` will create dest dir automatically

File: make/photon/portal/nginx.conf (hunk 7 above)
Comment (reasonerjt) [reply make/photon/portal/nginx.conf]: ok

File: make/photon/prepare/utils/nginx.py (hunk 9 above)
Comment (reasonerjt) [reply make/photon/prepare/utils/nginx.py]: I see they are parms to the func that mark the output file.

File: make/photon/jobservice/Dockerfile (hunk 4 above)
Comment (reasonerjt) [reply make/photon/jobservice/Dockerfile]: OK thanks.

File: make/photon/core/start.sh [hunk 15]
```diff
@@ -1,3 +1,3 @@
-#!/bin/sh
-sudo -E -u \#10000 "/harbor/harbor_core"
+#! sh
+./harbor/harbor_core
```
[the remark this answers is not in the document]
Comment (reasonerjt) [reply make/photon/core/start.sh]: Yes it is removed, so I'm resolving this comment.

File: make/photon/db/docker-entrypoint.sh [hunk 16]
```diff
@@ -23,95 +23,89 @@ file_env() {
...
-                                WARNING: No password has been set for the database.
-                                         This will allow anyone with access to the
-                                         Postgres port to access your database. In
-                                         Docker's default configuration, this is
-                                         effectively any other container on the same
-                                         system.
-                                         Use "-e POSTGRES_PASSWORD=password" to set
-                                         it in "docker run".
-                                ****************************************************
+# chown -R postgres:postgres $PGDATA
```
Comment (cadnce) [inline make/photon/db/docker-entrypoint.sh]: Remove if commented out

File: make/photon/db/Dockerfile [hunk 17]
```diff
@@ -18,15 +18,16 @@ RUN tdnf erase -y toybox && tdnf install -y util-linux net-tools
 
 VOLUME /var/lib/postgresql/data
 
-ADD ./make/photon/db/docker-entrypoint.sh /entrypoint.sh
-ADD ./make/photon/db/docker-healthcheck.sh /docker-healthcheck.sh
-RUN chmod u+x /entrypoint.sh /docker-healthcheck.sh
-ENTRYPOINT ["/entrypoint.sh"]
-HEALTHCHECK CMD ["/docker-healthcheck.sh"]
-
+COPY ./make/photon/db/docker-entrypoint.sh /entrypoint.sh
```
Comment (cadnce) [inline make/photon/db/Dockerfile]: Appreciate it's not your change that introduced this but this docker file is inconsistent with the others. It is copying to a different filename to where it was copied from.  E.g `docker-entrypoint.sh` becomes `entrypoint` 🤷‍♂️

File: make/prepare [hunk 18]
```diff
@@ -45,10 +45,11 @@ secret_dir=${data_path}/secret
 config_dir=$harbor_prepare_path/common/config
 
 # Run prepare script
-docker run --rm -v $input_dir:/input:z \
-                    -v $harbor_prepare_path:/compose_location:z \
-                    -v $config_dir:/config:z \
```
Comment (cadnce) [inline make/prepare]: The removal of this `:z` will break the ability to install harbor if you're running with selinux enabled.
 #8129

Comment (ninjadq) [reply make/prepare]: You are right. Thanks for mention

File: make/photon/prepare/utils/misc.py [hunk 19]
```diff
@@ -84,6 +84,22 @@ def prepare_config_dir(root, *name):
...
 
+def prepare_dir(root: str, *args, **kwargs) -> str:
+    gid, uid = kwargs.get('gid'), kwargs.get('uid')
+    absolute_path = Path(os.path.join(root, *args))
+    if absolute_path.is_file():
+        raise Exception('Path exists and the type is regular file')
+
+    absolute_path.mkdir(0o755, parents=True, exist_ok=True)
+
+    if not(gid is None or uid is None):
```
Comment (reasonerjt) [inline make/photon/prepare/utils/misc.py]: Could you update the code so when only `gid` or `uid` is set, the change can also be applied ?

File: make/photon/prepare/utils/redis.py [hunk 20]
```diff
@@ -0,0 +1,14 @@
+import os
+
+from g import data_dir, REDIS_UID, REDIS_GID
+from utils.misc import prepare_dir
+
+redis_data_path = os.path.join(data_dir, 'redis')
+
+def prepare_redis(config_dict):
+    prepare_dir(redis_data_path)
```
Comment (reasonerjt) [inline make/photon/prepare/utils/redis.py]: you can pass the REDIS_UID/REDIS_GID as parms to `prepare_dir`?

**Comment by cadnce** [comment]
:+1:  looks good, not sure what's going on with the co job though seems a network failure?
